File: 726-number-of-atoms/number-of-atoms.java
import java.util.*;

class Solution {
    public String countOfAtoms(String formula) {

        Stack<Map<String, Integer>> stack = new Stack<>();
        stack.push(new HashMap<>());

        int i = 0;
        int n = formula.length();

        while (i < n) {

            char ch = formula.charAt(i);

            if (ch == '(') {

                stack.push(new HashMap<>());
                i++;

            } else if (ch == ')') {

                Map<String, Integer> group = stack.pop();
                i++;

                int num = 0;

                while (i < n && Character.isDigit(formula.charAt(i))) {
                    num = num * 10 + (formula.charAt(i) - '0');
                    i++;
                }

                if (num == 0) {
                    num = 1;
                }

                Map<String, Integer> current = stack.peek();

                for (String atom : group.keySet()) {

                    int count = group.get(atom) * num;

                    current.put(
                        atom,
                        current.getOrDefault(atom, 0) + count
                    );
                }

            } else {

                StringBuilder atom = new StringBuilder();

                atom.append(formula.charAt(i));
                i++;

                while (
                    i < n &&
                    Character.isLowerCase(formula.charAt(i))
                ) {
                    atom.append(formula.charAt(i));
                    i++;
                }

                int num = 0;

                while (
                    i < n &&
                    Character.isDigit(formula.charAt(i))
                ) {
                    num = num * 10 + (formula.charAt(i) - '0');
                    i++;
                }

                if (num == 0) {
                    num = 1;
                }

                String name = atom.toString();

                stack.peek().put(
                    name,
                    stack.peek().getOrDefault(name, 0) + num
                );
            }
        }

        Map<String, Integer> count = stack.pop();

        List<String> atoms = new ArrayList<>(count.keySet());
        Collections.sort(atoms);

        StringBuilder ans = new StringBuilder();

        for (String atom : atoms) {

            ans.append(atom);

            if (count.get(atom) > 1) {
                ans.append(count.get(atom));
            }
        }

        return ans.toString();
    }
}

File: 726-number-of-atoms/number-of-atoms.py
class Solution:
    def countOfAtoms(self, formula: str) -> str:
        stack = [{}]
        i = 0
        n = len(formula)

        while i < n:

            if formula[i] == '(':
                stack.append({})
                i += 1

            elif formula[i] == ')':
                group = stack.pop()
                i += 1

                num = 0

                while i < n and formula[i].isdigit():
                    num = num * 10 + int(formula[i])
                    i += 1

                num = num or 1

                for atom, count in group.items():
                    stack[-1][atom] = (
                        stack[-1].get(atom, 0)
                        + count * num
                    )

            else:
                atom = formula[i]
                i += 1

                while i < n and formula[i].islower():
                    atom += formula[i]
                    i += 1

                num = 0

                while i < n and formula[i].isdigit():
                    num = num * 10 + int(formula[i])
                    i += 1

                num = num or 1

                stack[-1][atom] = (
                    stack[-1].get(atom, 0) + num
                )

        count = stack[0]

        ans = ""

        for atom in sorted(count):
            ans += atom

            if count[atom] > 1:
                ans += str(count[atom])

        return ans